### ISSUE TITLE
what are the validation constraints when a slot mentions then id for another class

### ISSUE THREAD

**Issue by turbomam** [microbiomedata/nmdc-schema#882]
write examples!

cc @aclum

**Comment by turbomam** [comment]
is the lack of referential integrity constraint desirable, because the creating instance (Activity) might be created before the created (DataObject?)

**Comment by aclum** [comment]
I would really like to get away from inserting objects a referential integrity constraint but I don't think we can yet. For example, with NEON the schema updates will happen in parallel to some of the workflow runs because of time constraints.

**Comment by turbomam** [comment]
Answered, and the answer turned out to be already written down.

`src/docs/v10-vs-v11-retrospective.md` has a section, "Challenges with the use of `structured_pattern`s in `slot_usage`s for approximating referential integrity, etc.", that works through this exact question using `associated_studies` as the example. It is both things this issue asked about, split:

- **Id pattern**: enforced in the schema by `structured_pattern` on a `slot_usage`, which is why those patterns appear in so many `slot_usage` blocks.
- **Referential integrity**: not enforced at write time. @aclum's comment above gives the reason, and it still holds. It is checked afterward instead, by `make check-references` (`src/scripts/check_references.py`) and by [refscan](https://github.com/microbiomedata/refscan) against the production database. Neither existed when this issue was filed.

The remaining work is that the explanation is unfindable where it sits and still says the tooling is being worked on. That is now https://github.com/microbiomedata/nmdc-schema/issues/3329 (Referential integrity vs id patterns: make the existing explanation findable and current), which also covers the same stale sentence in `src/docs/maintaining-the-schema.md`.

Closing in favor of https://github.com/microbiomedata/nmdc-schema/issues/3329.

**Comment by turbomam** [comment]
Correcting my closing comment above. It said referential integrity is "not enforced at write time" and attributed that to @aclum's 2023 comment. Both halves are wrong, and this is the first thing anyone searching for this question will find, so I do not want it left standing.

**It is enforced at write time.** In `nmdc-runtime`:

- `nmdc_runtime/api/db/mongo.py` defines `validate_json(in_docs, mdb, check_inter_document_references=False)`, which uses `refscan`'s `scan_outgoing_references`.
- `nmdc_runtime/api/endpoints/metadata.py` calls it with `check_inter_document_references=True` for `/metadata/json:submit`, and rejects broken references with HTTP 422.
- `nmdc_runtime/api/endpoints/workflows.py` does the same for workflow execution submissions.

**And the out-of-order case is handled rather than tolerated.** A violation is waived when the referenced document appears in the same JSON payload. From `metadata.py`: "check whether all documents referenced by the document exist, whether in the database or the same JSON payload." So submitting a study together with its biosamples passes; a genuinely dangling reference does not.

That is the answer to @aclum's 2023 concern about NEON schema updates running in parallel with workflow runs, rather than a decision to live without the constraint. I should not have written it as a deliberate choice.

The accurate split, and the answer to this issue's original question:

- **Shape of the id**: enforced in the schema, by `structured_pattern` on a `slot_usage`, after pattern materialization.
- **Existence of the referent**: not visible to schema validation at all, because that is per-document. Enforced by the runtime API on submission, and checkable after the fact with `refscan` and `make check-references`.

https://github.com/microbiomedata/nmdc-schema/pull/3332 now documents it this way in `src/docs/maintaining-the-schema.md`. Verified 2026-08-13 against the three runtime files named above.
